game-core/src/main/java/games/strategy/engine/data/unit/ability/ConvertUnitAbility.java

```diff
@@ -0,0 +1,64 @@
+package games.strategy.engine.data.unit.ability;
+
+import games.strategy.engine.data.UnitType;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.AccessLevel;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NonNull;
+import lombok.experimental.FieldDefaults;
+import lombok.experimental.NonFinal;
+
+/** Unit Ability to convert CombatUnitAbilities into other CombatUnitAbilities or into nothing */
+@Data
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+@Builder(toBuilder = true)
+public class ConvertUnitAbility {
+
+  enum Faction {
+    ALLIED,
+    ENEMY
+  }
+
+  @NonNull String name;
+  /** The unit types that have this ability */
+  @NonFinal Collection<UnitType> attachedUnitTypes;
+
+  /** Does this affect enemy and/or allied units */
+  @Builder.Default Collection<Faction> factions = List.of(Faction.ENEMY);
+
+  /** The unitAbility that will be converted */
+  @NonNull CombatUnitAbility from;
+
+  /** If to is null, then this just removes the "from" UnitAbility */
+  CombatUnitAbility to;
```

Comment:
Few comments here:
1. IMHO it's really useful to mark fields and parameters as `Nullable`. The IDE hints generated from those annotations really help to catch potential NPEs
2. Would a "no-unit-ability" ability make things simpler and avoid the need for a nullable combat unit ability? I could see that going both ways, I'm curious what your considerations were.
3. Is this potentially a collection, ie: a one-to-many relationship?

Reply:
> Would a "no-unit-ability" ability make things simpler and avoid the need for a nullable combat unit ability? I could see that going both ways, I'm curious what your considerations were.

That would probably work.  I don't currently see it causing an issue once the battle logic is worked out.  We might need to revisit this as that is worked on.

> Is this potentially a collection, ie: a one-to-many relationship?

For the built-in ConvertUnitAbilities, no.  But there might be a map that could make use of a one-to-many.  Do you want me to make it one-to-many?

Reply:
That might be a question for the forum. I could argue it both ways, better to do it now, but also would be a shame to over-complicate it with a one-to-many when it is always one-to-one. When we get to XML, we'll want that spec to land just once. That will be a time when we are far more locked in and the decision there then is more significant.

Reply:
I've added an EMPTY unit ability and set that as the default - https://github.com/triplea-game/triplea/pull/8646/commits/bb8bc0e7b7d2329facd7ddc1f0744040c3ab0cc9